runtime/runtime/src/config.rs

```diff
@@ -248,14 +316,32 @@ pub fn total_prepaid_exec_fees(
 pub fn total_deposit(actions: &[Action]) -> Result<Balance, IntegerOverflowError> {
     let mut total_balance: Balance = 0;
     for action in actions {
-        total_balance = safe_add_balance(total_balance, action.get_deposit_balance())?;
+        let action_balance;
+        if let Action::Delegate(signed_delegate_action) = action {
+            let actions = signed_delegate_action.delegate_action.get_actions();
+            action_balance = total_deposit(&actions)?;
+        } else {
+            action_balance = action.get_deposit_balance();
+        }
+        total_balance = safe_add_balance(total_balance, action_balance)?;
```

Comment:
Hm, this confused me a bit looking at it again, so probably a sign we need more comments. :) Sorry if we discussed it already but deposit handling might be the most critical part, as we could introduce various kinds of "free token" exploits if we are not careful with changes around the balance checker and refunds.

Unless I am mistaken, the code here implies that creating the `DelegateAction` on the relayer includes paying all token fees of inner actions, as part of paying `tx_cost` in `validate_transaction`. But if those actions fail, the refund actually goes to Alice. Is my understanding correct here?

I know you wanted to introduce `relayer_id` on receipts to enable refunds to the relayer. And I know the NEP makes it clear that Alice receives the refund. But it's not quite as clear that the relayer pays for gas **and token** fees of the inner actions. It actually makes sense to me to do it this way but we should be clear about that, as it affects security concerns for relayer implementations.

Regardless of what the answer is, can you please add a comment (here or somewhere else in the code) that explains who pays the deposit fees and who receives the refund?

Reply:
I added some comments. Please, check

Reply:
Perfect, thanks a lot for the clarifications and all the comments! It will be of great value for anyone navigating the nearcore code base.